### ISSUE TITLE
Improve search

### ISSUE THREAD

**Issue by arthurevans** [Polymer/polymer-element-catalog#165]
There are several flaws with the current search system:

- only searches by element name, not full text
- only matches element API docs, not guides
- would be nice to unify this search with the library search

**Comment by arthurevans** [comment]
Ideally, I think we'd do something like developer.android.com, where we provide autocomplete suggestions from the reference pages and guides, but if the user hits enter, do a custom search.

However, at the moment getting all elements and guides to show up would be a big win.

**Comment by marcus7777** [comment]
we can use http://open-elements.org/elements/lunr-js

**Comment by rictic** [comment]
Ooh, I really like the design of the `lunr-js` element. lunr.js in general has a very nice API, but I found that it was significantly slower than fullproof back when I was working on [sutsis](https://sutsis.nomei.la). I want to say it took something like twice as long to build the index, maybe more.

This was about a year ago though, and lunr may have improved in performance since then.

**Comment by addyosmani** [comment]
@rictic is it worth benchmarking the lunr library to see if index generation is anywhere near an acceptable level for the catalog? Agree the API seems pretty decent. Project has also stayed pretty active.

**Comment by rictic** [comment]
Yeah, fullproof is way less active, and the API is rougher, so if we can use lunr we probably should.

As I recall indexing ~6MB of data took ~10s on a Nexus 5 with lunr, and about half that with fullproof.

On reflection though, indexing time might not be as critical to the user flow for the element catalog as it was for sutsis's dictionary use case. For sutsis the user was mostly on mobile, and most likely wanted to look up one word and be done, so indexing speed was critical. For polymer documentation the user is more likely to be on a beefy machine, and waiting a bit for the indexing may not be as big a deal.

**Comment by marcus7777** [comment]
we can build the index on deploy and serve it as a file if it's too slow with `lunr-js` but as we are searching ~20Kb I think we're golden

**Comment by marcus7777** [comment]
`<lunr-js data="[[elements]]" search="{{q}}" output="{{_filteredElements}}"></lunr-js>` on app/elements/pages/page-browse.html

**Comment by marcus7777** [comment]
http://open-elements.org/browse?q=Material

**Comment by ebidel** [comment]
It's silly I can't search for "service worker" and the platinum elements dont come up :(

https://elements.polymer-project.org/browse?q=service&package=platinum-elements

**Comment by ebidel** [comment]
@blasten do you have time to take a look?
